gms/gossiper.cc

```diff
@@ -2535,7 +2535,8 @@ future<> gossiper::maybe_enable_features() {
         // gms::feature::enable should be run within seastar::async context
         return seastar::async([&features, &g] {
             std::set<std::string_view> features_v = boost::copy_range<std::set<std::string_view>>(features);
-            g._feature_service.enable(features_v);
+            auto features_e = g._feature_service.enable(features_v);
+            db::system_keyspace::persist_enabled_feature_info(features_e);
```

Comment:
Feature service can be thought of as a component which has an in-memory state and an on-disk state in order to survive restarts. Logically, the act of enabling a feature always consists of persisting the state on disk to guarantee monotonicity. Separating it outside results in decoherence of the component. It shouldn't be the responsibility of the gossiper to have this logic.

So what's the problem with feature service depending on system keyspace?

> This opens a way for removing qctx from {set|get}_scylla_local_param() set of helpers, but they are used by group0 helpers too, so it cannot be done right now :(

Whatever way this is solved, e.g. by passing some object directly to the helpers, you can also pass it as a dependency of the feature service.

Reply:
> So what's the problem with feature service depending on system keyspace?

Starting order. At the time sharded<feature_service>.start() is called the sharded<system_keyspace> is not yet (and cannot be) started.

One option is to pass sharded reference to feature_service constructor. This means that the feature service can call for uninitialized system_keyspace instances, so it needs checks for sys_ks.local_is_initialized(). Or state that feature service cannot/doesn't use system keyspace reference until certain event happens. Same on stop -- system keyspace will be stopped before feature service creating dangling (in aforementioned sense) reference.

Better option is to pass `system_keyspace&` into `feature_service::enable()`.

Reply:
Can we start feature_service later? If system keyspace does something to it when initializing, perhaps this should be moved into feature_service::start().

Reply:
> Can we start feature_service later?

The dependencies are

    system_keyspace --> {query_processor, database}
    query_processor --> {proxy}
    {database, proxy} --> feature_service

So `feature_service --> system_keyspace` completes the loop.

Not all those links are really needed construction-time. Like if we add `feature_service --> system_keyspace` dependency, it will stay unused until gossiper starts gossiping and enabling features. I admit something similar is true for e.g. `database --> feature_service` dependency, so we could construct database with feature service and start the f.-service later. But all these assumptions seem too fragile.


There's the third option btw. The `system_keyspace --> database` is here, but there's also `database --> { large_data_handler, compaction_manager, sstables_manager } --> system_keyspace` loop. It was cut ( f9b57df4718357eb7f1000a208620b78d7466b20 ) by system keyspace plugging itself to the database sub-service on boot and unplugging itself back on stop. Similarly it can plug/unplug itself to/from feature_service.

Reply:
I'm ok with plugging for the lack of better options.

Reply:
Or extract read-only part of feature_service and move the responsibility of manipulating and restoring state to a higher-level component initialized later. We have system_keyspace::enable_features_on_startup(), which belongs to that part. We could have system_keyspace::enable_feature(). But it seems like a move in the wrong direction, we shouldn't pack system_keyspace with responsibility for all components, but move the responsibility to individual components.

Reply:
The responsibility of manipulating the state is on gossiper and feature service is used purely as a notification engine that spreads this information around. If feature service was more than just that it would have the `feature_enabler` and would subscribe itself on gossiper events. Instead, the `feature_enabler` is gossiper component.

Reply:
Hm... Maybe the `feature_enabler` is the proper place to put system_keyspace reference on? Then it can kick system keyspace to preserve features and feature service to enable them

Reply:
It seems more appropriate than the gossiper. The latter should just do the gossipping.